pgx/src/heap_tuple/mod.rs

```diff
@@ -0,0 +1,337 @@
+        iter.for_each(|datum| {
+            nulls.push(datum.is_none());
+            datums.push(datum.unwrap_or(0));
+        });
+        if datums.len() != tupdesc.len() {
+            return Err(PgHeapTupleCreationError::IncorrectAttributeCount);
+        }
+
+        unsafe {
+            let formed_tuple =
+                pg_sys::heap_form_tuple(tupdesc.as_ptr(), datums.as_mut_ptr(), nulls.as_mut_ptr());
+
+            Ok(Self {
+                tuple: PgBox::<pg_sys::HeapTupleData, AllocatedByRust>::from_rust(formed_tuple),
+                tupdesc,
+            })
+        }
+    }
+
+    /// Creates a new [PgHeapTuple] from an opaque Datum that is should be a "composite" type.
+    ///
+    /// The Datum should be a pointer to a [pg_sys::HeapTupleHeader].  Typically, this will be used
+    /// in situations when working with SQL `ROW(...)` constructors, or a composite SQL type such as
+    ///
+    /// ```sql
+    /// CREATE TYPE my_composite AS (name text, age i32);
+    /// ```
+    ///
+    /// ## Safety
+    ///
+    /// This function is unsafe as we cannot guarantee that the provided Datum is a valid [pg_sys::HeapTupleHeader]
+    /// pointer.
+    pub unsafe fn from_composite_datum(composite: pg_sys::Datum) -> Self {
+        let htup_header =
+            pg_sys::pg_detoast_datum(composite as *mut pg_sys::varlena) as pg_sys::HeapTupleHeader;
+        let tup_type = crate::heap_tuple_header_get_type_id(htup_header);
+        let tup_typmod = crate::heap_tuple_header_get_typmod(htup_header);
+        let tupdesc = pg_sys::lookup_rowtype_tupdesc(tup_type, tup_typmod);
+
+        let mut data = PgBox::<pg_sys::HeapTupleData>::alloc0();
+
+        data.t_len = crate::heap_tuple_header_get_datum_length(htup_header) as u32;
+        data.t_data = htup_header;
+
+        Self {
+            tuple: data,
+            tupdesc: PgTupleDesc::from_pg(tupdesc),
+        }
+    }
+
+    /// Given the name for an attribute in this [PgHeapTuple], change its value.
+    ///
+    /// Attribute names are case sensitive.
+    ///
+    /// ## Errors
+    ///
+    /// - return [TryFromDatumError::NoSuchAttributeName] if the attribute does not exist
+    /// - return [TryFromDatumError::IncompatibleTypes] if the Rust type of the `value` is not
+    /// compatible with the attribute's Postgres type
+    pub fn set_by_name<T: IntoDatum>(&mut self, attname: &str, value: T) -> FromDatumResult<()> {
+        match self.get_attribute_by_name(attname) {
+            None => Err(TryFromDatumError::NoSuchAttributeName(attname.to_string())),
+            Some((attnum, _)) => self.set_by_index(attnum, value),
+        }
+    }
+
+    /// Given the index for an attribute in this [PgHeapTuple], change its value.
+    ///
+    /// Attribute numbers start at 1, not 0.
+    ///
+    /// ## Errors
+    /// - return [TryFromDatumError::NoSuchAttributeIndex] if the attribute does not exist
```

Comment:
```suggestion
    /// - return [pgx::datum::::from::TryFromDatumError::NoSuchAttributeIndex] if the attribute does not exist
```

This is a dead link: 

```rust

warning: unresolved link to `TryFromDatumError::NoSuchAttributeIndex`
   --> pgx/src/heap_tuple/mod.rs:172:19
    |
172 |     /// - return [TryFromDatumError::NoSuchAttributeIndex] if the attribute does not exist
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the enum `TryFromDatumError` has no variant or associated item named `NoSuchAttributeIndex`

```